scripts/mk_all.lean

```diff
@@ -0,0 +1,108 @@
+/-
+Copyright (c) 2024 Yaël Dillies, Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Damiano Testa
+-/
+import Cli.Basic
+import Lean.Util.Path
+import Lake.CLI.Main
+
+/-!
+# Script to create a file importing all files from a folder
+
+This file declares a command to gather all Lean files from a folder into a single Lean file.
+-/
+
+open Lean System.FilePath
+
+/-- `getAll git ml` takes all `.lean` files in the dir `ml` (recursing into sub-dirs) and
+returns the string
+```
+import file₁
+...
+import fileₙ
+```
+The input `git` is a `Bool`ean flag:
+* `true` means that the command uses `git ls-files` to find the relevant files;
+* `false` means that the command recursively scans all dirs searching for `.lean` files.
+-/
+def getAll (git : Bool) (ml : String) : IO String := do
+  let ml.lean := addExtension ⟨ml⟩ "lean"  -- for example, `Mathlib.lean`
+  let allModules : Array System.FilePath ← (do
+    if git then
+      let mlDir := ml.push pathSeparator   -- for example, `Mathlib/`
+      let allLean ← IO.Process.run { cmd := "git", args := #["ls-files", mlDir ++ "*.lean"] }
+      return (((allLean.dropRightWhile (· == '\n')).splitOn "\n").map (⟨·⟩)).toArray
+    else do
+      let all ← walkDir ml
+      return all.filter (·.extension == some "lean"))
+  let files := (allModules.erase ml.lean).qsort (·.toString < ·.toString)
+  let withImport ← files.mapM fun f => do
+    -- this check is helpful in case the `git` option is on and a local file has been removed
+    if ← pathExists f then
+      return "import " ++ (← moduleNameOfFileName f none).toString
+    else return ""
+  return ("\n".intercalate (withImport.filter (· != "")).toList).push '\n'
+
+open Lake in
+/-- `getLeanLibs` returns the names (as an `Array` of `String`s) of all the libraries
+on which the current project depends.
+If the current project is `mathlib`, then it excludes the libraries `Cache` and `LongestPole` and
+it includes `Mathlib/Tactic`. -/
+def getLeanLibs : IO (Array String) := do
+  let (elanInstall?, leanInstall?, lakeInstall?) ← findInstall?
+  let config ← MonadError.runEIO <| mkLoadConfig { elanInstall?, leanInstall?, lakeInstall? }
+  let ws ← MonadError.runEIO (MainM.runLogIO (loadWorkspace config)).toEIO
+  let package := ws.root
+  let libs := (package.leanLibs.map (·.name)).map (·.toString)
+  return if package.name == `mathlib then
+    libs.erase "Cache" |>.erase "LongestPole" |>.push ("Mathlib".push pathSeparator ++ "Tactic")
```

Comment:
I suppose this is okay for now. Probably we would want some sort of mechanism in Lake allowing us to differentiate build dependencies and dev dependencies.

Reply:
Good point!  I added a `TODO` in the doc-module highlighting this point.